starknet-rpc-test/src/lib.rs

```diff
@@ -30,50 +57,72 @@ pub enum ExecutionStrategy {
     Wasm,
 }
 
+#[derive(Error, Debug)]
+pub enum TestError {
+    #[error("No free ports")]
+    NoFreePorts,
+}
+
 impl Drop for MadaraClient {
     fn drop(&mut self) {
-        if let Err(e) = self.process.kill() {
-            eprintln!("Could not kill Madara process: {}", e)
+        self.process.kill().expect("Cannot kill process");
+    }
+}
+
+fn get_free_port() -> Result<u16, TestError> {
+    for port in MIN_PORT..=MAX_PORT {
+        if let Ok(listener) = TcpListener::bind(("127.0.0.1", port)) {
+            return Ok(listener.local_addr().expect("No local addr").port());
         }
+        // otherwise port is occupied
     }
+
+    Err(TestError::NoFreePorts)
 }
 
 impl MadaraClient {
-    fn init(execution: ExecutionStrategy) -> Self {
-        let madara_path = Path::new("../target/release/madara");
-        assert!(
-            madara_path.exists(),
-            "could not find the madara binary at `{}`",
-            madara_path.to_str().expect("madara_path must be a valid path")
-        );
-
-        let child_handle = Command::new(madara_path.to_str().unwrap())
+    async fn init(execution: ExecutionStrategy) -> Result<Self, TestError> {
+        // we keep the reference, otherwise the mutex unlocks immediately
+        let _mutex_guard = BACKGROUND_MADARA_MUTEX.lock().await;
+
+        let free_port = get_free_port()?;
+
+        let root = Path::new("../");
+        std::env::set_current_dir(&root).expect("Failed to change working directory");
```

Comment:
Will this continue for the whole test execution?
Because in those tests we will probably want to access other files by doing things like "../compiled_contracts/".